### PR TITLE
[SYCL][NVPTX] Set default fdiv and sqrt for llvm.fpbuiltin

### PR DESCRIPTION
**Pull request by MrSidims**
AltMathLibrary is lacking implementation for llvm.fpbuiltin intrinsics
for NVPTX target. This patch adds type-dependent mapping for
llvm.fpbuiltin.fdiv with max-error > 2.0 and llvm.fpbuiltin.sqrt with
max-error > 1.0 on nvvm intrinsics:
fp32 scalar @llvm.fpbuiltin.fdiv -> @llvm.nvvm.div.approx.f
fp32 scalar @llvm.fpbuiltin.sqrt -> @llvm.nvvm.sqrt.approx.f

vector or non-fp32 scalar llvm.fpbuiltin.fdiv -> fdiv
vector or non-fp32 scalar llvm.fpbuiltin.sqrt -> llvm.sqrt

Additionally it maps max-error=0.5 fpbuiltin.fadd, fpbuiltin.fsub.
fpbuiltin.fmul, fpbuiltin.fdiv, fpbuiltin.frem, fpbuiltin.sqrt and
fpbuiltin.ldexp intrinsic functions of LLVM's math operations or
https://llvm.org/docs/LangRef.html#standard-c-c-library-intrinsics

TODO in future patches:
- add preservation of debug info in FPBuiltinFnSelection;
- moved tests from CodeGen to Transform
- move pass to new pass manager

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
@zahiraam FYI this PR should fix c-math LIT failure on CUDA in https://github.com/intel/llvm/pull/15836
@npmiller @Naghasan please take a look at the PR or assign somebody familiar with CUDA/NVPTX as I'm quite fresh in this topic.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 1]
```diff
@@ -154,6 +197,14 @@ static bool selectFnForFPBuiltinCalls(const TargetLibraryInfo &TLI,
     }
   }
 
+  // We don't have implementation for CUDA approximate precision builtins.
+  // Lets map them on NVPTX intrinsics. If no appropriate intrinsics are known
+  // - skip to replaceWithAltMathFunction.
+  if (T.isNVPTX() && BuiltinCall.getRequiredAccuracy().value() == 3.0) {
+    if (replaceWithNVPTXCalls(BuiltinCall))
```
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:204]: Should we somehow encode "skip"/"fallback" to alt math functions part into the name?

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 2]
```diff
@@ -106,6 +107,48 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
   return true;
 }
 
+static bool replaceWithNVPTXCalls(FPBuiltinIntrinsic &BuiltinCall) {
```
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:110]: We only call this function if requested precision is exactly `3.0`, but the name is quite generic. I think that it worth adding a comment right before the function to better specify its intent/expected use case of this function in case it will be extended in the future

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-3.0.ll [hunk 3]
```diff
@@ -0,0 +1,81 @@
+; RUN: opt -fpbuiltin-fn-selection -S < %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-3.0.ll:1]: I would prefer those tests to be in `Transform` instead, but I'm not against adding a single new test into `CodeGen` simply because they should either all be in the right location, or none of them (i.e. the move should be done as a separate PR).

**Review by AlexeySachkov — COMMENTED** [review]
The code itself looks good to me, just a few naming/documentation comments.
I'm also unfamiliar with CUDA, so it would be great to hear feedback from @intel/llvm-reviewers-cuda

**Comment by zahiraam** [comment]
> @zahiraam FYI this PR should fix c-math LIT failure on CUDA in #15836 @npmiller @Naghasan please take a look at the PR or assign somebody familiar with CUDA/NVPTX as I'm quite fresh in this topic.

This change makes sense to me. I will wait until this PR is merged in, then will rebase and test again and hopefully the `cmath_test.cpp` fail will get fixed.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 4]
```diff
@@ -106,6 +107,48 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
...
+  // For vector fpbuiltins for NVPTX target we don't have nvvm intrinsics, use
+  // standart for LLVM math operations. Also nvvm fdiv and sqrt intrisics
+  // support only float type.
+  switch (BuiltinCall.getIntrinsicID()) {
+  case Intrinsic::fpbuiltin_fdiv:
+    if (Type->isVectorTy() || !Type->getScalarType()->isFloatTy())
+      return replaceWithLLVMIR(BuiltinCall);
+    Replacement =
+        IRBuilder.CreateIntrinsic(Type,
+                                  FMF.isFast()
```
Comment (frasercrmck) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:128]: The NVPTX backend only selects the "FTZ" variants when:

``` cpp
bool NVPTXTargetLowering::useF32FTZ(const MachineFunction &MF) const {
  return MF.getDenormalMode(APFloat::IEEEsingle()).Output ==
         DenormalMode::PreserveSign;
}
```

Selecting it here based on the fast-math flags might not be correct.

Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp]: Thanks! I've removed the check for fast math flag and ftz generation leaving the backend to decide, what PTX instruction to use.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 1 above)
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp]: Not sure, from my perspective the code here is not complex and speaks for itself :)

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-3.0.ll (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-3.0.ll:1]: Totally agree, lets keep the tests in one place for now. If necessary we will move some of them to `Transform `dir

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp]: My point was not in the complexity, but in the expectations. 

Let's say you read the this specific function, i.e. you start top-down to understand high-level structure first. Just by looking at the name, you may expect that this specific step will introduce some NVPTX-specific intrinsics, but in fact, it will do some other form of lowering which could be surprising for you unless you also went through the function itself.

Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp]: Sure, renamed and clarified in the comment

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 5]
```diff
@@ -106,6 +107,42 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
...
+// attribute to the appropriate nvvm approximate intrinsics if it's possible.
+static bool replaceWithApproxNVPTXCalls(FPBuiltinIntrinsic &BuiltinCall) {
+  IRBuilder<> IRBuilder(&BuiltinCall);
+  SmallVector<Value *> Args(BuiltinCall.args());
+  Value *Replacement = nullptr;
+  auto *Type = BuiltinCall.getType();
+  // For now only add lowering for fdiv and sqrt. Yet nvvm intrinsics have
+  // approximate variants for sin, cos, exp2 and log2.
+  // For vector fpbuiltins for NVPTX target we don't have nvvm intrinsics, use
+  // standart for LLVM math operations. Also nvvm fdiv and sqrt intrisics
```
Comment (frasercrmck) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:120]: `standart` looks like a typo. I'm not sure what the comment is trying to say, sorry.

**Review by frasercrmck — COMMENTED** [review]
The PR description will need updating to reflect the FTZ changes

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 5 above)
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:123]: Thanks! I was referring to https://llvm.org/docs/LangRef.html#standard-c-c-library-intrinsics. Fixed

**Review by frasercrmck — APPROVED** [review]

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 6]
```diff
@@ -106,6 +107,46 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
...
+  auto *Type = BuiltinCall.getType();
+  // For now only add lowering for fdiv and sqrt. Yet nvvm intrinsics have
+  // approximate variants for sin, cos, exp2 and log2.
+  // For vector fpbuiltins for NVPTX target we don't have nvvm intrinsics,
+  // fallback to instruction or standard C/C++ library LLVM intrinsic. Also
+  // nvvm fdiv and sqrt intrisics support only float type, so fallback in this
+  // case as well.
+  switch (BuiltinCall.getIntrinsicID()) {
+  case Intrinsic::fpbuiltin_fdiv:
+    if (Type->isVectorTy() || !Type->getScalarType()->isFloatTy())
```
Comment (jchlanda) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:128]: This is a bit of a nit-picking, so feel free to ignore. Since both cases are identical, bar the intrinsic ID, I'd be tempted to only set the ID in the body of `case`, and do the type check/create intrinsic afterwards.

**Review by jchlanda — APPROVED** [review]

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 6 above)
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:128]: I see your point and I agree that it would be better - I was thinking to do it like you have said initially, but then I've intentionally left it inside the switch case. The reason is (see the comment above: "For now only add lowering for fdiv and sqrt."): there are other nvvm intrinsics, that might or might not support vectors and/or doubles. For example looking at NVPTXIntrinsics.td I see, that there are both `lg2.approx.f32` and `lg2.approx.f64` overloads of `lg2` intrinsic. In this patch I'm solving a narrow issue, but I can imagine this switch case being expanded in the future, so for now those ifs are moved inside.

**Comment by MrSidims** [comment]
@intel/dpcpp-tools-reviewers , @AlexeySachkov please take a look

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 7]
```diff
@@ -154,6 +195,14 @@ static bool selectFnForFPBuiltinCalls(const TargetLibraryInfo &TLI,
     }
   }
 
+  // We don't have implementation for CUDA approximate precision builtins.
+  // Lets map them on NVPTX intrinsics. If no appropriate intrinsics are known
+  // - skip to replaceWithAltMathFunction.
+  if (T.isNVPTX() && BuiltinCall.getRequiredAccuracy().value() == 3.0) {
```
Comment (jcranmer-intel) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:201]: Looking at the PTX ISA documentation (https://docs.nvidia.com/cuda/parallel-thread-execution/#floating-point-instructions-div), the accuracy varies based on the operation. For example, div has 2 ULP error, and sqrt has 1 ULP error (if I haven't botched my conversion characteristics). The other functions you mention in the above comment have much, much higher ULPs in general.

So I think it makes sense to let the check for replacement vary the needed accuracy on a per-function basis.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 8]
```diff
@@ -106,6 +107,51 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
...
+  auto *Type = BuiltinCall.getType();
+  // For now only add lowering for fdiv and sqrt. Yet nvvm intrinsics have
+  // approximate variants for sin, cos, exp2 and log2.
+  // For vector fpbuiltins for NVPTX target we don't have nvvm intrinsics,
+  // fallback to instruction or standard C/C++ library LLVM intrinsic. Also
+  // nvvm fdiv and sqrt intrisics support only float type, so fallback in this
+  // case as well.
+  switch (BuiltinCall.getIntrinsicID()) {
+  case Intrinsic::fpbuiltin_fdiv:
+    if (Accuracy.value() != 2.5)
```
Comment (MrSidims) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:129]: There was actually a bug - for fdiv I should have checked if max-error is 2.5, not 3.0

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 7 above)
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp]: Thanks! I have applied the comment in this way: check here for `== 3.0` was replaced to `!= 0.5` and accuracy parameter was moved to be a function parameter. Inside of the function I've added a check of accuracy for `fdiv == 2.5` (or for `sqrt == 3.0`), then try to replace `fpbuiltin` with `approx` nvvm intrinsic, otherwise function does nothing expecting `fpbuiltin` intrinsics to be handled by alt math library. May be it makes sense to check, if accuracy for `fdiv >= 2.0` (for `sqrt >= 1.0`) and then to replace with such `fpbuiltin`s with `approx` nvvm intrinsics,, but I'm not sure if it's a good general solution, may be it's better to get a test failure later on during approximate/fast builtins feature development, then silently replace them here.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 9]
```diff
@@ -154,6 +200,14 @@ static bool selectFnForFPBuiltinCalls(const TargetLibraryInfo &TLI,
     }
   }
 
+  // We don't have implementation for CUDA approximate precision builtins.
```
Comment (maksimsab) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:203]: nit: in case of upstreaming it would be unclear who are "we".

**Review by maksimsab — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 8 above)
Comment (jcranmer-intel) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:129]: The accuracy check here should be a >= check, not a != check.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 10]
```diff
@@ -106,6 +107,51 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
...
+  case Intrinsic::fpbuiltin_fdiv:
+    if (Accuracy.value() != 2.5)
+      return false;
+    if (Type->isVectorTy() || !Type->getScalarType()->isFloatTy())
+      return replaceWithLLVMIR(BuiltinCall);
+    Replacement =
+        IRBuilder.CreateIntrinsic(Type, Intrinsic::nvvm_div_approx_f, Args);
+    break;
+  case Intrinsic::fpbuiltin_sqrt:
+    if (Accuracy.value() != 3.0)
```
Comment (jcranmer-intel) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:137]: Ditto.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 11]
```diff
@@ -154,6 +200,14 @@ static bool selectFnForFPBuiltinCalls(const TargetLibraryInfo &TLI,
     }
   }
 
+  // We don't have implementation for CUDA approximate precision builtins.
+  // Lets map them on NVPTX intrinsics. If no appropriate intrinsics are known
+  // - skip to replaceWithAltMathFunction.
+  if (T.isNVPTX() && BuiltinCall.getRequiredAccuracy().value() != 0.5)
```
Comment (jcranmer-intel) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:206]: This should be a > check (not >=), I believe.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 8 above)
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:129]: To double check, that we are on the same page, in case if max-error >= 2.5 like this it will fall back to alt math impl. I guess we would want an opposite behavior, when if max-error >= 2.5 -> replace with approx nvvm intrinsic, right?

Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:129]: Actually, you are right, with an extension, that it should be just `>` check (I still need to catch `2.5` and `3.0` values to resolve issue in https://github.com/intel/llvm/pull/15836 )

Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:129]: Actually, let me rewrite the patch.

**Comment by MrSidims** [comment]
@jchlanda @frasercrmck @maksimsab hi, following conversation with @jcranmer-intel I've significantly changed the patch since your approval, may I ask you to take another look?

What have changed:
Added lowering for 0.5 aka precise fpbuiltin intrinsics as AltMathLibrary don't handle them for NVPTX/CUDA. Lowering for approx nvvm is now dependent on max-error value and follows https://docs.nvidia.com/cuda/parallel-thread-execution/#floating-point-instructions , so for example for fdiv it starts from 2.0 ULP. Note, that range between 0.5 and 2.0 ULP for fdiv is not covered by this patch.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 12]
```diff
@@ -106,6 +107,51 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
...
+    Replacement =
+        IRBuilder.CreateIntrinsic(Type, Intrinsic::nvvm_div_approx_f, Args);
+    break;
+  case Intrinsic::fpbuiltin_sqrt:
+    if (Accuracy.value() < 1.0)
+      return false;
+    if (Type->isVectorTy() || !Type->getScalarType()->isFloatTy())
+      return replaceWithLLVMIR(BuiltinCall);
+    Replacement = IRBuilder.CreateIntrinsic(
+        BuiltinCall.getType(), Intrinsic::nvvm_sqrt_approx_f, Args);
```
Comment (jchlanda) [inline llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:142]:
```
        Type, Intrinsic::nvvm_sqrt_approx_f, Args);
```

To align with the case above.

**Review by jchlanda — APPROVED** [review]

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 8 above)
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp]: Align with nvptx spec. Also added lowering for precise fdiv/sqrt

**Review by maksimsab — APPROVED** [review]

**Review by frasercrmck — COMMENTED** [review]
Generally looks good, just some questions.

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp [hunk 13]
```diff
@@ -106,6 +107,51 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
   return true;
 }
 
+// This function lowers llvm.fpbuiltin. intrinsic functions with max-error
+// attribute to the appropriate nvvm approximate intrinsics if it's possible.
+// If it's not possible - fallback to instruction or standard C/C++ library LLVM
+// intrinsic.
+static bool
+replaceWithApproxNVPTXCallsOrFallback(FPBuiltinIntrinsic &BuiltinCall,
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:115]: No, it doesn't copy debug info, the entire pass is not doing this. Let me fix it in follow up patch, for now I'd like to unblock https://github.com/intel/llvm/pull/15836

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll [hunk 14]
```diff
@@ -0,0 +1,72 @@
+; RUN: opt -fpbuiltin-fn-selection -S < %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll:1]: > Is the pass intentionally limited to the legacy pass manager?

I don't know :) I'm just fixing things around for now.

> update_test_checks.py

I might be missing something, but it's output was not really helpful:

```
diff --git a/llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll b/llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll
index 51b18b4e3bab..7aff496e8391 100644
--- a/llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll
+++ b/llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll
@@ -1,4 +1,5 @@
-; RUN: opt -fpbuiltin-fn-selection -S < %s | FileCheck %s
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --tool ./build/bin/opt --version 5
+; RUN: opt -passes=fpbuiltin-fn-selection -S < %s | FileCheck %s

 target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
 target triple = "nvptx64-nvidia-cuda"
@@ -7,7 +8,7 @@ target triple = "nvptx64-nvidia-cuda"
 ; CHECK: %{{.*}} = call float @llvm.nvvm.div.approx.f(float %{{.*}}, float %{{.*}})
 ; CHECK: %{{.*}} = fdiv <2 x float> %{{.*}}, %{{.*}}
 define void @test_fdiv(float %d1, <2 x float> %v2d1,
-                       float %d2, <2 x float> %v2d2) {
+  float %d2, <2 x float> %v2d2) {
 entry:
   %t0 = call float @llvm.fpbuiltin.fdiv.f32(float %d1, float %d2) #0
   %t1 = call <2 x float> @llvm.fpbuiltin.fdiv.v2f32(<2 x float> %v2d1, <2 x float> %v2d2) #0
@@ -18,7 +19,7 @@ entry:
 ; CHECK: %{{.*}} = call fast float @llvm.nvvm.div.approx.f(float %{{.*}}, float %{{.*}})
 ; CHECK: %{{.*}} = fdiv fast <2 x float> %{{.*}}, %{{.*}}
 define void @test_fdiv_fast(float %d1, <2 x float> %v2d1,
-                            float %d2, <2 x float> %v2d2) {
+  float %d2, <2 x float> %v2d2) {
 entry:
   %t0 = call fast float @llvm.fpbuiltin.fdiv.f32(float %d1, float %d2) #0
   %t1 = call fast <2 x float> @llvm.fpbuiltin.fdiv.v2f32(<2 x float> %v2d1, <2 x float> %v2d2) #0
@@ -32,7 +33,7 @@ declare <2 x float> @llvm.fpbuiltin.fdiv.v2f32(<2 x float>, <2 x float>)
 ; CHECK: %{{.*}} = fdiv double %{{.*}}, %{{.*}}
 ; CHECK: %{{.*}} = fdiv <2 x double> %{{.*}}, %{{.*}}
 define void @test_fdiv_double(double %d1, <2 x double> %v2d1,
-                              double %d2, <2 x double> %v2d2) {
+  double %d2, <2 x double> %v2d2) {
 entry:
   %t0 = call double @llvm.fpbuiltin.fdiv.f64(double %d1, double %d2) #0
   %t1 = call <2 x double> @llvm.fpbuiltin.fdiv.v2f64(<2 x double> %v2d1, <2 x double> %v2d2) #0

```

[the remark this answers is not in the document]
Comment (frasercrmck) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll:1]: It might be because `-passes=fpbuiltin-fn-selection` doesn't work without an entry for the pass with the new PM. I don't know if the script works with the legacy syntax or not. Did you change the RUN line before running the script?

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (frasercrmck) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:115]: Fair enough, it can wait!

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll (hunk 14 above)
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll:1]: > Did you change the RUN line before running the script?

Oh, indeed, I've just tried to change it see if the pass is registered - and indeed it's not. But with this changed reverted - again there no changes in CHECK lines. May be when the pass is registered it would also fix the script.

[the remark this answers is not in the document]
Comment (frasercrmck) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll:1]: Yeah, perhaps. It would be best to update all the other tests at the same time anyway, so it can wait.

**Review by frasercrmck — APPROVED** [review]

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll (hunk 14 above)
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-approx.ll:1]: Along with moving them from CodeGen to Transform

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-0.5.ll [hunk 15]
```diff
@@ -0,0 +1,219 @@
+; RUN: opt -fpbuiltin-fn-selection -S < %s | FileCheck %s
+
+; Several functions for "sycl" and "cuda" requires "0.5" accuracy levels,
```
Comment (bader) [inline llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-0.5.ll:3]:
```suggestion
; Several functions for SYCL and CUDA requires "0.5" accuracy levels,
```

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-0.5.ll [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+; Test if these fpbuiltins could be replaced by equivalaent IR operations
+; or llvm builtins.
```
Comment (bader) [inline llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-0.5.ll:5]:
```suggestion
; or LLVM builtins.
```

File: llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-0.5.ll (hunk 15 above)
Comment (MrSidims) [reply llvm/test/CodeGen/NVPTX/fp-builtin-intrinsics-nvvm-max-error-0.5.ll]: applied

**Review by jcranmer-intel — COMMENTED** [review]
One last nit:
Can you add a test that uses a rather large fpbuiltin-max-error, say 10 or 100? Just to cover the non-equality cases for the various if statements.

Other than that, this looks good.

**Review by jcranmer-intel — APPROVED** [review]

**Comment by MrSidims** [comment]
@intel/llvm-gatekeepers please help with merge

File: llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp:115]: Added to https://github.com/intel/llvm/pull/16847
